### PR TITLE
Fix Issue 17270 - std.experimental.Final fails on pointers

### PR DESCRIPTION
**Pull request by Bolpat**

### CODE REVIEW

File: std/experimental/typecons.d [hunk 1]
```diff
@@ -1024,3 +1024,10 @@ pure nothrow @safe unittest
     static assert(!__traits(compiles, arr ~= 4));
     assert((arr ~ 4) == [1, 2, 3, 4]);
 }
+
```
Comment (wilzbach) [inline std/experimental/typecons.d]: Usually the link to the bug report is added as comment to the unittest, so that future developers know about this.

File: std/experimental/typecons.d [hunk 2]
```diff
@@ -1024,3 +1024,10 @@ pure nothrow @safe unittest
...
+pure nothrow @nogc @system unittest
+{
+    int i = 1;
+    Final!(int*) fp = &i;
+    assert(*fp == 1);
```
Comment (wilzbach) [inline std/experimental/typecons.d]: How about adding more tests for pointers here? 
- assignment to the `fp` should fail 
- copying `fp` should be possible

**Comment by Bolpat** [comment]
Looking at another aspect. What is assignment? We have
```
@disable void opUnary(string op : "--")();
@disable void opUnary(string op : "++")();
```
but is it really assignment *in general?* One could argue against because `++` and `--` can be overloaded. For a type where `++` is `const`, it wouldn't change the object at all and is no assignment.

File: std/experimental/typecons.d (hunk 2 above)
Comment (quickfur) [inline std/experimental/typecons.d]: Hmm. What about checking that `typeof(*fp) == int`?

**Comment by quickfur** [comment]
@Bolpat  I'm not sure what to make of disabling `++` and `--`.  What was the original reasoning behind this?

**Comment by Bolpat** [comment]
@quickfur It's not. Should be better discussed in the [General group](https://forum.dlang.org/group/general), right?

**Comment by quickfur** [comment]
You're right.

File: std/experimental/typecons.d (hunk 2 above)
Comment (Bolpat) [inline std/experimental/typecons.d]: It feels somewhat ridiculous, but ... okay.

Comment (quickfur) [inline std/experimental/typecons.d]: The concern is that it should not be some other wrapper type masquerading as int. I'd also check that `i` has the same value as `*fp` after the assignment.

**Comment by quickfur** [comment]
Also, how would pointer arithmetic interact with `Final!(int*)`? Does it do the expected thing? I.e., the unittest should check this.

I'm asking this not to be nitpicky, but because after playing around a little with `Final`, I have discovered some surprising behaviour, that's a side-effect of the way it's implemented. For example:
````
Final!(int[]) arr;
arr.length++;  // correctly rejected
arr.length = 10; // OK!
arr.ptr = null; // correctly rejected
arr.ptr++; // correctly rejected
````

There may be other such surprising corner cases lurking in there, that a more thorough unittest might uncover.

**Comment by Bolpat** [comment]
Arrays are another spacial case. We should have a separate unit test for them. But yes, good point. I haven't thought about that in my [forum post](https://forum.dlang.org/post/ggucqgdfdvrwewdkfqwt@forum.dlang.org). I'll add them to the PR.

File: std/experimental/typecons.d [hunk 3]
```diff
@@ -1024,3 +1024,34 @@ pure nothrow @safe unittest
...
+    assert(*p == 2);
+}
+
+pure nothrow @system unittest
+{
+    Final!(int[]) arr;
+    static assert(!__traits(compiles,
+        arr.length++
+    ));
+    arr.length = 10;
```
Comment (quickfur) [inline std/experimental/typecons.d]: This should not work; allowing modification of array length shouldn't be allowed. Or at least, if it's allowed, then modifying `.ptr` should also be allowed.  Since this is a bug, it should not be put into a unittest, because whatever is in a unittest implies that it's supposed to be allowed.

I would write the `arr.length` test as a not-compilable test, but commented out with a comment indicating the bug number.  But this is actually beyond the scope of this PR.  I would split out the array-related unittest into a separate PR, actually, so that we can merge the pointer fix now rather than later. Fixing array behaviour is something else that needs to be discussed, but shouldn't block the original intent of this PR.

**Comment by quickfur** [comment]
ping @Bolpat

File: std/experimental/typecons.d (hunk 3 above)
Comment (Bolpat) [inline std/experimental/typecons.d]: Exactly.

**Comment by Bolpat** [comment]
I'm working on a new implementation which addresses the semantics on structs.

**Comment by quickfur** [comment]
This PR is starting to suffer from mission creep.

Let's just merge this now, and address struct support in a separate PR.
